drivers/flash/flash_stm32l5x.c

```diff
@@ -0,0 +1,381 @@
+		return -EIO;
+	}
+
+	/* Check that no Flash memory operation is ongoing */
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if ((regs->OPTR & FLASH_OPTR_DBANK) == FLASH_OPTR_DBANK) {
+		bool bank_swap;
+		/* Check whether bank1/2 are swapped */
+		bank_swap =
+		((regs->OPTR & FLASH_OPTR_SWAP_BANK) == FLASH_OPTR_SWAP_BANK);
+
+		if ((offset < (FLASH_SIZE / 2)) && !bank_swap) {
+			/* The pages to be erased is in bank 1 */
+			regs->NSCR &= ~FLASH_NSCR_NSBKER_Msk;
+			page = offset / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 1", page);
+		} else if ((offset >= BANK2_OFFSET) && bank_swap) {
+			/* The pages to be erased is in bank 1 */
+			regs->NSCR &= ~FLASH_NSCR_NSBKER_Msk;
+			page = (offset - BANK2_OFFSET) / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 1", page);
+		} else if ((offset < (FLASH_SIZE / 2)) && bank_swap) {
+			/* The pages to be erased is in bank 2 */
+			regs->NSCR |= FLASH_NSCR_NSBKER;
+			page = offset / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 2", page);
+		} else if ((offset >= BANK2_OFFSET) && !bank_swap) {
+			/* The pages to be erased is in bank 2 */
+			regs->NSCR |= FLASH_NSCR_NSBKER;
+			page = (offset - BANK2_OFFSET) / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 2", page);
+		} else {
+			LOG_ERR("Offset %d does not exist", offset);
+			return -EINVAL;
+		}
+	} else {
+		page = offset / FLASH_PAGE_SIZE;
+		LOG_DBG("Erase page %d\n", page);
+	}
+
+	/* Set the NSPER bit and select the page you wish to erase */
+	regs->NSCR |= FLASH_NSCR_NSPER;
+	regs->NSCR &= ~FLASH_NSCR_NSPNB_Msk;
+	regs->NSCR |= (page << FLASH_NSCR_NSPNB_Pos);
+
+	/* Set the NSSTRT bit */
+	regs->NSCR |= FLASH_NSCR_NSSTRT;
+
+	/* flush the register write */
+	tmp = regs->NSCR;
+
+	/* Wait for the NSBSY bit */
+	rc = flash_stm32_wait_flash_idle(dev);
+
+	if ((regs->OPTR & FLASH_OPTR_DBANK) == FLASH_OPTR_DBANK) {
+		regs->NSCR &= ~(FLASH_NSCR_NSPER | FLASH_NSCR_NSBKER);
+	} else {
+		regs->NSCR &= ~(FLASH_NSCR_NSPER);
+	}
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(const struct device *dev,
+				 unsigned int offset,
+				 unsigned int len)
+{
+	unsigned int address = offset;
+	int rc = 0;
+	bool icache_enabled = LL_ICACHE_IsEnabled();
+
+	if (icache_enabled) {
+		/* Disable icache */
+		stm32l5x_icache_disable();
+		stm32l5x_icache_clear_status();
+	}
+
+	for (; address <= offset + len - 1 ; address += FLASH_PAGE_SIZE) {
+		rc = erase_page(dev, address);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	if (icache_enabled) {
+		/* Enable icache */
+		stm32l5x_icache_enable();
+		/* Invalidate icache */
+		rc = stm32l5x_icache_invalidate();
+
+		stm32l5x_icache_clear_status();
+	}
+
+	return rc;
+}
+
+int flash_stm32_write_range(const struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int i, rc = 0;
+	bool icache_enabled = LL_ICACHE_IsEnabled();
+
+	if (icache_enabled) {
+		/* Disable icache */
+		stm32l5x_icache_disable();
```

Comment:
status return by stm32l5x_icache_disable() should be checked

Reply:
As said, invalidate should be necessary only for Erase operation.

Reply:
> As said, invalidate should be necessary only for Erase operation.

Invalidation should be done after erase and after write. During write and erase operation i-cache should be disabled. Disabling i-cache will trigger the invalidation process automatically.

![image](https://user-images.githubusercontent.com/78502494/127660181-cc4e25fd-597d-4ae3-a946-568bbd470e31.png)

So the investigations i have done and the observations/results match with this. i will raise an updated PR shortly.

Reply:
@ikmdani ,
great, lets follow the Reference Manual recommendation (even if it is only a recommendation)